### DISCUSSION TITLE
Sql server + Authorization

### DISCUSSION THREAD

**Discussion by mesfandi** [elsa-workflows/elsa-core#1046] [Q&A] [answered]
Hi,
I am wondering if I can integrate Elsa with my (or any other) authentication and Authorization system. (I am using Identity Server for my application). Or is there any way that I add a column to the workflow database and store the user Id there?
I want each user can access only the workflows they created.

my second question is how I can use Sql server instead of sqlite? I saw there exists Elsa.Persistence.EntityFramework.SqlServer in NuGet. However, I am not sure how I can use that.
Thank you

**Comment by mesfandi** [comment]
Also, is there any way to implement Authentication and Authorization for the API calls, so, each user can only call the APIs they created?

**Comment by sfmskywalker** [comment, marked as answer]
You should be able to integrate Elsa with any authentication system you like.

If you want each user to access only the workflows they created, it means you have a User & Permission strategy in place.
So what you might do is create a table that associates a workflow with a user, so that you application can check if a user is the owner of the workflow they are trying to access.

Regarding your second question, I just updated [the documentation](https://elsa-workflows.github.io/elsa-core/docs/next/installation/installing-persistence#sql-server) that answers your questions:

o use SQL Server as the EF Core provider, add the Elsa.Persistence.EntityFramework.SqlServer package:

```
dotnet add package Elsa.Persistence.EntityFramework.SqlServer
```

Then update your Startup class as follows:

```csharp
using Elsa.Persistence.EntityFramework.SqlServer

public void ConfigureServices(IServiceCollection services)
{
    services.AddElsa(options => options.UseEntityFrameworkPersistence(ef => ef.UseSqlServer("Server=local;Database=Elsa")));
}
```

**Reply by mesfandi** [reply]
Thank you very much for the answer. 
actually my question is about the dashboard. I could not find the API it consumes to show the workflows. 

I have another question in this regard. I have another DbContext and when I add sql server, it throws me an exception:
`System.InvalidOperationException: 'The DbContextOptions passed to the ApplicationDbContext constructor must be a DbContextOptions<ApplicationDbContext>. When registering multiple DbContext types, make sure that the constructor for each context type has a DbContextOptions<TContext> parameter `
 
I modified the code and made it generic 
`.AddElsa(elsa => elsa.UseEntityFrameworkPersistence<ElsaContext>(ef =>ef.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")))`

It does not work. 
could you please help me with this one.
Thank you.

**Reply by sfmskywalker** [reply]
You got it.

Sure. So both `ElsaContext` and `ApplicationDbContext`  now have a constructor that accepts `DbContextOptions<ElsaContext>` and `DbContextOptions<ApplicationDbContext>`?

If so, then perhaps you can share your project with me so I can take a look?

**Reply by mesfandi** [reply]
Thank you so much. It worked. I had a small problem with my constructor. 
As my final question, would you please let me know which API is consumed by the Dashboard and How I can Modify that. 
I want to show only workflows that the user created. I used my authentication and user table as you mentioned. However, the Dashboard gets all the workflows. I need to filter the workflows based on the users and send it to the dashboard. 
I really appreciate your help and your nice ELSA.

**Reply by sfmskywalker** [reply]
@mesfandi The Dashboard consumes [this API endpoint](https://github.com/elsa-workflows/elsa-core/blob/master/src/server/Elsa.Server.Api/Endpoints/WorkflowDefinitions/List.cs) for listing workflow definitions. There's currently no way for you to easily filter out workflows created by users, but if you create a feature request in the form of an issue, we can discuss the details there to see what we'd need to change to allow for that.

**Reply by AizenSousuke** [reply]
> You should be able to integrate Elsa with any authentication system you like.
> 
> If you want each user to access only the workflows they created, it means you have a User & Permission strategy in place. So what you might do is create a table that associates a workflow with a user, so that you application can check if a user is the owner of the workflow they are trying to access.
> 
> Regarding your second question, I just updated [the documentation](https://elsa-workflows.github.io/elsa-core/docs/next/installation/installing-persistence#sql-server) that answers your questions:
> 
> o use SQL Server as the EF Core provider, add the Elsa.Persistence.EntityFramework.SqlServer package:
> 
> ```
> dotnet add package Elsa.Persistence.EntityFramework.SqlServer
> ```
> 
> Then update your Startup class as follows:
> 
> ```cs
> using Elsa.Persistence.EntityFramework.SqlServer
> 
> public void ConfigureServices(IServiceCollection services)
> {
>     services.AddElsa(options => options.UseEntityFrameworkPersistence(ef => ef.UseSqlServer("Server=local;Database=Elsa")));
> }
> ```

For the automatic migration using the above code, I just have to create the Database "Elsa"? I've tried it and there's an error about not migrating and cannot find the tables.

**Comment by asaadhoussam** [comment]
hi
Thank you for the clarification and useful explanation. I have a question, please:
What will the configuration be like when I use an external identity server such as the Identity Server4.0, that controls several sites, one of which is the Elsa workflow3.0? Will I keep using the local Identity in Elsa3.0? Or can it be dispensed with? Or is there another way?
Thank you all
